core/src/main/java/org/springframework/security/jackson2/UnmodifiableSetDeserializer.java

```diff
@@ -50,10 +50,10 @@ public Set deserialize(JsonParser jp, DeserializationContext ctxt) throws IOExce
 				Iterator<JsonNode> nodeIterator = arrayNode.iterator();
 				while (nodeIterator.hasNext()) {
 					JsonNode elementNode = nodeIterator.next();
-					resultSet.add(mapper.readValue(elementNode.toString(), Object.class));
+					resultSet.add(mapper.readValue(elementNode.traverse(mapper), Object.class));
```

Comment:
... just by chance noticed this. May have been changed since then, but method `mapper.treeToValue()` would perhaps be best method to use here. It does translate to very similar call so it's not a big deal either way, and code above works fine.